syntax = "proto3";

message Payload {
  enum Action {
    CREATE_PROJECT = 0;
    CREATE_TASK = 1;
    PROGRESS_TASK = 2;
    EDIT_TASK = 3;
    ADD_USER = 4;
  }

  // An action tells the handler which function
  // the payload requested to use.
  Action action = 1;

  // Time the transaction was sent.
  uint64 timestamp = 2;
  
  // All the data neccessitated by the actions
  // defined in the enum.
  CreateProjectAction create_project = 3;
  CreateTaskAction create_task = 4;
  ProgressTaskAction progress_task = 5;
  EditTaskAction edit_task = 6;
  AddUserAction add_user = 7;
}

message CreateProjectAction {
  //project name
  string project_name = 1;

}

message CreateTaskAction {
  string project_name = 1;
  string task_name = 2;
  string description = 3;
}

message ProgressTaskAction {
  string project_name = 1;
  string task_name = 2;
}

message EditTaskAction {
  string project_name = 1;
  string task_name = 2;
  string description = 3;
}

message AddUserAction {
  string project_name = 1;
  string public_key = 2;
}